File: src/comfort_cues/app_state.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from pathlib import Path
import json
import os


@dataclass(slots=True)
class AppState:
    app_enabled: bool = True
    launch_to_tray: bool = False
    ui_language: str = "en"


def default_data_root() -> Path:
    return default_app_state_path().parent


def default_app_state_path() -> Path:
    if os.name == "nt":
        base = Path(os.environ.get("APPDATA", Path.home() / "AppData" / "Roaming"))
        return base / "Comfort Cues" / "app-state.json"
    return Path.home() / ".comfort-cues" / "app-state.json"


def load_app_state(path: Path) -> AppState:
    if not path.exists():
        return AppState()

    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError, TypeError):
        return AppState()

    return AppState(
        app_enabled=bool(payload.get("app_enabled", True)),
        launch_to_tray=bool(payload.get("launch_to_tray", False)),
        ui_language="zh" if str(payload.get("ui_language", "en")).strip().lower() == "zh" else "en",
    )


def save_app_state(path: Path, state: AppState) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(asdict(state), ensure_ascii=False, indent=2) + "\n", encoding="utf-8")

File: native/src/app_state.cpp
#include "app_state.h"

#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QJsonDocument>
#include <QJsonObject>

QString defaultDataRoot()
{
    return QFileInfo(defaultAppStatePath()).absolutePath();
}

QString defaultAppStatePath()
{
    QString base = qEnvironmentVariable("APPDATA");
    if (base.isEmpty()) {
        base = QDir::homePath() + "/AppData/Roaming";
    }
    return QDir::toNativeSeparators(base + "/Comfort Cues/app-state.json");
}

AppState loadAppState(const QString &path)
{
    QFile file(path);
    AppState state;
    if (!file.exists() || !file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        return state;
    }

    const QJsonDocument document = QJsonDocument::fromJson(file.readAll());
    const QJsonObject object = document.object();
    state.appEnabled = object.value("app_enabled").toBool(true);
    state.launchToTray = object.value("launch_to_tray").toBool(false);
    state.uiLanguage = object.value("ui_language").toString("en").trimmed().toLower();
    if (state.uiLanguage != "zh") {
        state.uiLanguage = "en";
    }
    return state;
}

bool saveAppState(const QString &path, const AppState &state)
{
    QDir().mkpath(QFileInfo(path).absolutePath());
    QFile file(path);
    if (!file.open(QIODevice::WriteOnly | QIODevice::Text | QIODevice::Truncate)) {
        return false;
    }

    QJsonObject object;
    object.insert("app_enabled", state.appEnabled);
    object.insert("launch_to_tray", state.launchToTray);
    object.insert("ui_language", state.uiLanguage);

    QByteArray payload = QJsonDocument(object).toJson(QJsonDocument::Indented);
    if (!payload.endsWith('\n')) {
        payload.append('\n');
    }
    file.write(payload);
    return true;
}
